test/runtests.jl

```diff
@@ -0,0 +1,71 @@
+###-----------------------------------------------------------------------------
+### Copyright (C) 2022- Cursor Insight
+###
+### All rights reserved.
+###-----------------------------------------------------------------------------
+
+###=============================================================================
+### Imports
+###=============================================================================
+
+using Test
+
+using Patterns: @pattern,  @with_pattern
+
+###=============================================================================
+### Tests
+###=============================================================================
+
+@testset "Pattern" begin
+    @with_pattern function f end
+
+    @test f isa Function
+    @test length(methods(f)) == 1 # there is a default wrapper
+    @test hasmethod(f, Tuple{Vararg{Symbol}})
+
+    @pattern function f(:x)
+        return "x"
+    end
+
+    @test length(methods(f)) == 2
+    @test hasmethod(f, Tuple{Val{:x}})
+    @test f(:x) == "x"
+
+    @pattern function f(:y)
+        return 'y'
+    end
+
+    @test length(methods(f)) == 3
+    @test hasmethod(f, Tuple{Val{:y}})
+    @test f(:y) == 'y'
+
+    @pattern function f(:z, num::Integer)
+        return num
+    end
+
+    @test length(methods(f)) == 4
+    @test hasmethod(f, Tuple{Val{:y}, Integer})
+    @test f(:z, 42) == 42
+
+    @pattern function f(:a, :b, :c)
+        return "abc"
+    end
+
+    @test length(methods(f)) == 5
+    @test hasmethod(f, Tuple{Val{:a}, Val{:b}, Val{:c}})
+    @test f(:a, :b, :c) == "abc"
+
+    # couldn't create pattern with return type assertion
+    try
+        # TODO this is how you can catch exception from macro
+        quote
+            @pattern function f(:y)::Int
+                return 3
+            end
+        end |> eval
```

Comment:
With `@eval`, you can get rid of the `quote`. Better yet, you can use `@macroexpand` instead of `@eval`, which does not evaluate the return value of the macro (i.e., does not actually define the function), but still runs the macro itself, so assertions actually happen.

And it gets even better! In Julia 1.7, assertions thrown from a macro while running `@macroexpand` do not get wrapped in a `LoadError`, so there is no need to use the convoluted test below, one can simply write `@test_throws AssertionError @macroexpand @pattern ....` In Julia 1.6 and older, you still have to go through the loops below.